dotnet/src/Connectors/Connectors.Memory.Redis/RedisVectorRecordStore.cs

```diff
@@ -0,0 +1,399 @@
+    private readonly string _keyJsonPropertyName;
+
+    /// <summary>An array of the names of all the data properties that are part of the redis payload, i.e. all properties except the key and vector properties.</summary>
+    private readonly string[] _dataPropertyNames;
+
+    /// <summary>An array of the names of all the data and vector properties that are part of the redis payload.</summary>
+    private readonly string[] _dataAndVectorPropertyNames;
+
+    /// <summary>The mapper to use when mapping between the consumer data model and the redis record.</summary>
+    private readonly IVectorStoreRecordMapper<TRecord, (string Key, JsonNode Node)> _mapper;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="RedisVectorRecordStore{TRecord}"/> class.
+    /// </summary>
+    /// <param name="database">The redis database to read/write records from.</param>
+    /// <param name="options">Optional configuration options for this class.</param>
+    /// <exception cref="ArgumentNullException">Throw when parameters are invalid.</exception>
+    public RedisVectorRecordStore(IDatabase database, RedisVectorRecordStoreOptions<TRecord>? options)
+    {
+        // Verify.
+        Verify.NotNull(database);
+
+        // Assign.
+        this._database = database;
+        this._options = options ?? new RedisVectorRecordStoreOptions<TRecord>();
+
+        // Enumerate public properties using configuration or attributes.
+        (PropertyInfo keyProperty, List<PropertyInfo> dataProperties, List<PropertyInfo> vectorProperties) properties;
+        if (this._options.VectorStoreRecordDefinition is not null)
+        {
+            properties = VectorStoreRecordPropertyReader.FindProperties(typeof(TRecord), this._options.VectorStoreRecordDefinition, supportsMultipleVectors: true);
+        }
+        else
+        {
+            properties = VectorStoreRecordPropertyReader.FindProperties(typeof(TRecord), supportsMultipleVectors: true);
+        }
+
+        // Validate property types and store for later use.
+        VectorStoreRecordPropertyReader.VerifyPropertyTypes([properties.keyProperty], s_supportedKeyTypes, "Key");
+        VectorStoreRecordPropertyReader.VerifyPropertyTypes(properties.vectorProperties, s_supportedVectorTypes, "Vector");
+
+        this._keyPropertyInfo = properties.keyProperty;
+        this._keyJsonPropertyName = VectorStoreRecordPropertyReader.GetSerializedPropertyName(this._keyPropertyInfo);
+
+        this._dataPropertyNames = properties
+            .dataProperties
+            .Select(VectorStoreRecordPropertyReader.GetSerializedPropertyName)
+            .ToArray();
+
+        this._dataAndVectorPropertyNames = this._dataPropertyNames
+            .Concat(properties.vectorProperties.Select(VectorStoreRecordPropertyReader.GetSerializedPropertyName))
+            .ToArray();
+
+        // Assign Mapper.
+        if (this._options.MapperType == RedisRecordMapperType.JsonNodeCustomMapper)
```

Comment:
I actually do have a third mapper option in mind for redis, where the input/output to/from it is a json string.  That way implementers can control their own json parsing/serialization.
That said, it is certainly possible to just check if a mapper is set and using it if it's not null, instead of using the enum.
We can always throw if more than one is set.
I do however like the explicitness of having an enum, but I suppose it's only really required if we have multiple built in options, since you wouldn't be able to choose the right one otherwise.
I do also like having a consistent experience across all the connectors, so if one has an enum, because it has multiple built in options, but another does not, it can be confusing.
I'm thinking we should table this decision a bit while we build more connectors, to see how more of them differ, and if we find that most are just a simple default + override setup, it'll be an easy decision to remove the enums.

Reply:
Yeah that sounds fine. Will changing the enum be a breaking change?

Reply:
Adding items shouldn't be, although in the strictest sense it can break users of it if they encounter a new unknown value. 
 That said, we are the only real consumers of this.
Removing the enum will be breaking.  We should definitely make this decision before we merge the feature branch.